dBFT: recalculate proposed Merkle state if needed

AnnaShaleva opened the pull request:
Close #140. Close #60. Port #141 with some changes.

In this PR:
* Ensure all header's fields that affect miner's state calculations are properly set in `(*DBFT).Prepare`. A part of #140.
* Recalculate Merkle state depending on updated dBFT context information in case of CV. Close #140.
* Optimize accepted block insertion into chain using state calculated during PrepareRequest-level block verification (port the #124). Close #60.

Tested on privnet with transactions, works as expected. @chenquanyu, @nicolegys, welcome to review and test.

AnnaShaleva committed:
dbft: always use in-turn difficulty for miner's task preparation

Provide valid block difficulty to the miner so that miner's state for
the proposed block is based on this difficulty. In good case (no CVs
during block acceptance process) dBFT won't recalculate block's state.
In bad case (at least one CV present) dBFT will recalculate block state
based on the updated difficulty. Ref. #140.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft, core: fix method comments

Make them follow go comment conventions.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
core: reduce block verification errors logging

The only error log we need is the log of inability to retrieve
parent block by hash. All other errors are logged on the dBFT service
level.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: refactor WithProcessBlock dBFT callback

Follow DRY and rename variables to match their meaning.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: simplify ToEthBlock handler

`(b *Block) WithWithdrawals` is able to handle  both nil and empty
withdrawals.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: recalculate block state after CV

Perform block state recalculation if CV happened since miner's proposal
may become outdated. Close #140.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: use precalculated state (if available) to store block

The idea of this commit is to keep block's receipts got after
PrepareRequest level block verification and use them while submitting
the resulting block to the chain in order to save some time on receipts
recalculation. Close #60.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

roman-khimov commented on consensus/dbft/dbft.go:
```diff
@@ -448,10 +423,34 @@ func New(config *params.DBFTConfig, _ ethdb.Database) (*DBFT, error) {
+
+					log.Crit("failed to process proposal state",
+						"err", err,
+						"rlp", hex.EncodeToString(w.Bytes()))
```
Can be hunderds of KBs.

AnnaShaleva replied:
Fixed, added only several fields instead.

roman-khimov commented on consensus/dbft/dbft.go:
```diff
@@ -448,10 +423,34 @@ func New(config *params.DBFTConfig, _ ethdb.Database) (*DBFT, error) {
+					w := bytes.NewBuffer(nil)
+					_ = ethBlock.EncodeRLP(w)
+
+					log.Crit("failed to process proposal state",
```
"proposed state"? "state from proposal"?

AnnaShaleva replied:
Fixed.

roman-khimov commented on consensus/dbft/dbft.go:
```diff
@@ -448,10 +423,34 @@ func New(config *params.DBFTConfig, _ ethdb.Database) (*DBFT, error) {
 			// Fill in only proposal and receipts, transactions will be properly
 			// set from context later in SetTransactionHashes callback.
 			req.SealingProposal = c.sealingProposal
-
```
:eyes:

AnnaShaleva replied:
Moved to a separate comment since this change is intentional.

roman-khimov commented on consensus/dbft/dbft.go:
```diff
@@ -582,6 +581,38 @@ func New(config *params.DBFTConfig, _ ethdb.Database) (*DBFT, error) {
+	binary.BigEndian.PutUint64(h.Nonce[:], uint64(ctx.PrimaryIndex))
+
+	h.Difficulty = c.getDifficulty(int(ctx.PrimaryIndex), uint64(ctx.BlockIndex))
+
```
MixDigest?

AnnaShaleva replied:
MixDigest is set during PrepareRequest construction and CNs will agree on it during consensus, so it shouldn't be here anymore. Ref. https://github.com/bane-labs/go-ethereum/pull/121#discussion_r1473416855.

roman-khimov commented on consensus/dbft/block.go:
```diff
@@ -3,6 +3,7 @@ package dbft
 import (
 	"errors"
 	"fmt"
+	"github.com/ethereum/go-ethereum/core/state"
```
goimports

AnnaShaleva replied:
Fixed.

AnnaShaleva committed:
dbft: adjust comment and fix codestyle

No functional changes, just unify the logical block of code and adjust
comment.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>
